File: mobile/lib/services/user_service.dart
import 'package:get/get.dart' hide Response;
import '../models/models.dart';
import 'api_service.dart';

/// 用户服务
class UserService {
  ApiService get _api => ApiService.to;

  /// 获取用户列表
  Future<List<User>> getUsers({
    String? keyword,
    String? departmentId,
    bool? isActive,
    int page = 1,
    int pageSize = 20,
  }) async {
    final params = <String, dynamic>{
      'page': page,
      'pageSize': pageSize,
    };
    if (keyword != null && keyword.isNotEmpty) params['keyword'] = keyword;
    if (departmentId != null) params['departmentId'] = departmentId;
    if (isActive != null) params['isActive'] = isActive;

    final resp = await _api.get('/users', queryParameters: params);
    final data = resp.data;
    if (data['code'] == 200) {
      final items = data['data']?['items'] ?? data['data'] ?? [];
      if (items is List) {
        return items.map((e) => User.fromJson(e as Map<String, dynamic>)).toList();
      }
    }
    return [];
  }

  /// 获取用户详情
  Future<User?> getUserById(String id) async {
    final resp = await _api.get('/users/$id');
    final data = resp.data;
    if (data['code'] == 200 && data['data'] != null) {
      return User.fromJson(data['data'] as Map<String, dynamic>);
    }
    return null;
  }

  /// 创建用户
  Future<User?> createUser(CreateUserRequest req) async {
    final resp = await _api.post('/users', data: req.toJson());
    final data = resp.data;
    if (data['code'] == 200 && data['data'] != null) {
      return User.fromJson(data['data'] as Map<String, dynamic>);
    } else {
      throw ApiException(
        code: data['code'] ?? -1,
        message: data['message'] ?? '创建用户失败',
      );
    }
  }

  /// 更新用户
  Future<bool> updateUser(String id, UpdateUserRequest req) async {
    final resp = await _api.put('/users/$id', data: req.toJson());
    final data = resp.data;
    if (data['code'] == 200) return true;
    throw ApiException(
      code: data['code'] ?? -1,
      message: data['message'] ?? '更新用户失败',
    );
  }

  /// 删除用户
  Future<bool> deleteUser(String id) async {
    final resp = await _api.delete('/users/$id');
    final data = resp.data;
    if (data['code'] == 200) return true;
    throw ApiException(
      code: data['code'] ?? -1,
      message: data['message'] ?? '删除用户失败',
    );
  }

  /// 修改用户状态
  Future<bool> updateUserStatus(String id, bool isActive) async {
    final resp = await _api.patch('/users/$id/status', data: {'isActive': isActive});
    final data = resp.data;
    if (data['code'] == 200) return true;
    throw ApiException(
      code: data['code'] ?? -1,
      message: data['message'] ?? '修改状态失败',
    );
  }

  /// 分配用户角色
  Future<bool> updateUserRoles(String id, List<String> roleIds) async {
    final resp = await _api.put('/users/$id/roles', data: {'roleIds': roleIds});
    final data = resp.data;
    if (data['code'] == 200) return true;
    throw ApiException(
      code: data['code'] ?? -1,
      message: data['message'] ?? '分配角色失败',
    );
  }

  /// 重置密码
  Future<bool> resetPassword(String id) async {
    final resp = await _api.put('/users/$id/password');
    final data = resp.data;
    if (data['code'] == 200) return true;
    throw ApiException(
      code: data['code'] ?? -1,
      message: data['message'] ?? '重置密码失败',
    );
  }

  /// 获取我的权限列表
  Future<List<String>> getMyPermissions() async {
    final resp = await _api.get('/users/permissions/my');
    final data = resp.data;
    if (data['code'] == 200 && data['data'] != null) {
      if (data['data'] is List) {
        return List<String>.from(data['data']);
      }
    }
    return [];
  }
}

File: backend-new/src/LimsAuth.Api/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using LimsAuth.Api.Data;
using LimsAuth.Api.Models;
using LimsAuth.Api.Models.Entities;

namespace LimsAuth.Api.Services;

public interface IUserService
{
    Task<ApiResponse<PagedResponse<UserListItemDto>>> GetListAsync(UserQueryRequest query);
    Task<ApiResponse<UserDetailDto>> GetByIdAsync(Guid id);
    Task<ApiResponse<UserListItemDto>> CreateAsync(CreateUserRequest request);
    Task<ApiResponse<UserListItemDto>> UpdateAsync(Guid id, UpdateUserRequest request);
    Task<ApiResponse> DeleteAsync(Guid id);
    Task<ApiResponse> UpdateRolesAsync(Guid id, UpdateUserRolesRequest request);
    Task<ApiResponse> ResetPasswordAsync(Guid id, ResetPasswordRequest request);
    Task<ApiResponse<List<RoleBriefDto>>> GetRolesAsync(Guid id);
    Task<ApiResponse<List<PermissionBriefDto>>> GetPermissionsAsync(Guid id);
}

public class UserService : IUserService
{
    private readonly AppDbContext _db;

    public UserService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<ApiResponse<PagedResponse<UserListItemDto>>> GetListAsync(UserQueryRequest query)
    {
        var q = _db.SysUsers
            .Include(u => u.UserRoles).ThenInclude(ur => ur.Role)
            .Include(u => u.MainDepartment)
            .Where(u => u.IsDeleted == 0)
            .AsQueryable();

        if (!string.IsNullOrEmpty(query.Keyword))
            q = q.Where(u => u.Username.Contains(query.Keyword) || u.RealName.Contains(query.Keyword));
        if (query.DepartmentId.HasValue)
            q = q.Where(u => u.MainDepartmentId == query.DepartmentId);
        if (query.IsActive.HasValue)
            q = q.Where(u => u.Status == (query.IsActive.Value ? 1 : 0));

        var total = await q.CountAsync();
        var items = await q
            .OrderByDescending(u => u.CreatedAt)
            .Skip((query.Page - 1) * query.PageSize)
            .Take(query.PageSize)
            .Select(u => new UserListItemDto
            {
                Id = u.Id,
                Username = u.Username,
                RealName = u.RealName,
                EmployeeNo = u.EmployeeNo,
                Gender = u.Gender,
                Mobile = u.Mobile,
                Email = u.Email,
                MainInstitutionName = u.MainInstitution != null ? u.MainInstitution.Name : null,
                MainDepartmentName = u.MainDepartment != null ? u.MainDepartment.Name : null,
                UserType = u.UserType,
                Status = u.Status,
                Avatar = u.Avatar,
                LastLoginTime = u.LastLoginTime,
                CreatedAt = u.CreatedAt,
                Roles = u.UserRoles.Select(ur => new RoleBriefDto { Id = ur.Role.Id, Code = ur.Role.Code, Name = ur.Role.Name }).ToList()
            })
            .ToListAsync();

        return ApiResponse<PagedResponse<UserListItemDto>>.Success(new PagedResponse<UserListItemDto>
        {
            Items = items, Total = total, Page = query.Page, PageSize = query.PageSize
        });
    }

    public async Task<ApiResponse<UserDetailDto>> GetByIdAsync(Guid id)
    {
        var u = await _db.SysUsers
            .Include(u => u.UserRoles).ThenInclude(ur => ur.Role).ThenInclude(r => r.RolePermissions).ThenInclude(rp => rp.Permission)
            .Include(u => u.MainDepartment)
            .FirstOrDefaultAsync(u => u.Id == id && u.IsDeleted == 0);

        if (u == null) return ApiResponse<UserDetailDto>.Error(404, "用户不存在");

        var perms = u.UserRoles
            .SelectMany(ur => ur.Role.RolePermissions)
            .Select(rp => rp.Permission)
            .Distinct()
            .Select(p => new PermissionBriefDto { Id = p.Id, Code = p.Code, Name = p.Name })
            .ToList();

        return ApiResponse<UserDetailDto>.Success(new UserDetailDto
        {
            Id = u.Id,
            Username = u.Username,
            RealName = u.RealName,
            EmployeeNo = u.EmployeeNo,
            Gender = u.Gender,
            Mobile = u.Mobile,
            Email = u.Email,
            MainDepartmentName = u.MainDepartment?.Name,
            UserType = u.UserType,
            Status = u.Status,
            Avatar = u.Avatar,
            LastLoginTime = u.LastLoginTime,
            CreatedAt = u.CreatedAt,
            Roles = u.UserRoles.Select(ur => new RoleBriefDto { Id = ur.Role.Id, Code = ur.Role.Code, Name = ur.Role.Name }).ToList(),
            Permissions = perms
        });
    }

    public async Task<ApiResponse<UserListItemDto>> CreateAsync(CreateUserRequest request)
    {
        if (await _db.SysUsers.AnyAsync(u => u.Username == request.Username && u.IsDeleted == 0))
            return ApiResponse<UserListItemDto>.Error(400, "用户名已存在");

        var user = new SysUser
        {
            Id = Guid.NewGuid(),
            Username = request.Username,
            Password = BCrypt.Net.BCrypt.HashPassword(request.Password),
            RealName = request.RealName,
            EmployeeNo = request.EmployeeNo,
            Gender = request.Gender,
            Mobile = request.Mobile,
            Email = request.Email,
            MainInstitutionId = request.MainInstitutionId,
            MainDepartmentId = request.MainDepartmentId,
            UserType = request.UserType,
            Status = request.Status,
            CreatedAt = DateTime.UtcNow
        };

        _db.SysUsers.Add(user);

        if (request.RoleIds.Any())
        {
            foreach (var roleId in request.RoleIds)
            {
                _db.SysUserRoles.Add(new SysUserRole { UserId = user.Id, RoleId = roleId });
            }
        }

        await _db.SaveChangesAsync();

        return ApiResponse<UserListItemDto>.Success(new UserListItemDto
        {
            Id = user.Id,
            Username = user.Username,
            RealName = user.RealName,
            Status = user.Status,
            CreatedAt = user.CreatedAt
        }, "用户创建成功");
    }

    public async Task<ApiResponse<UserListItemDto>> UpdateAsync(Guid id, UpdateUserRequest request)
    {
        var user = await _db.SysUsers.FirstOrDefaultAsync(u => u.Id == id && u.IsDeleted == 0);
        if (user == null) return ApiResponse<UserListItemDto>.Error(404, "用户不存在");

        if (!string.IsNullOrEmpty(request.RealName)) user.RealName = request.RealName;
        if (request.Gender.HasValue) user.Gender = request.Gender.Value;
        if (!string.IsNullOrEmpty(request.Mobile)) user.Mobile = request.Mobile;
        if (!string.IsNullOrEmpty(request.Email)) user.Email = request.Email;
        if (request.MainInstitutionId.HasValue) user.MainInstitutionId = request.MainInstitutionId;
        if (request.MainDepartmentId.HasValue) user.MainDepartmentId = request.MainDepartmentId;
        if (!string.IsNullOrEmpty(request.UserType)) user.UserType = request.UserType;
        if (request.Status.HasValue) user.Status = request.Status.Value;
        user.UpdatedAt = DateTime.UtcNow;

        await _db.SaveChangesAsync();

        return ApiResponse<UserListItemDto>.Success(new UserListItemDto
        {
            Id = user.Id,
            Username = user.Username,
            RealName = user.RealName,
            Status = user.Status
        }, "用户更新成功");
    }

    public async Task<ApiResponse> DeleteAsync(Guid id)
    {
        var user = await _db.SysUsers.FirstOrDefaultAsync(u => u.Id == id && u.IsDeleted == 0);
        if (user == null) return ApiResponse.Error(404, "用户不存在");
        user.IsDeleted = 1;
        user.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        return ApiResponse.Success("用户删除成功");
    }

    public async Task<ApiResponse> UpdateRolesAsync(Guid id, UpdateUserRolesRequest request)
    {
        var user = await _db.SysUsers.Include(u => u.UserRoles).FirstOrDefaultAsync(u => u.Id == id && u.IsDeleted == 0);
        if (user == null) return ApiResponse.Error(404, "用户不存在");

        _db.SysUserRoles.RemoveRange(user.UserRoles);
        foreach (var roleId in request.RoleIds)
        {
            _db.SysUserRoles.Add(new SysUserRole { UserId = user.Id, RoleId = roleId });
        }
        await _db.SaveChangesAsync();
        return ApiResponse.Success("角色分配成功");
    }

    public async Task<ApiResponse> ResetPasswordAsync(Guid id, ResetPasswordRequest request)
    {
        var user = await _db.SysUsers.FirstOrDefaultAsync(u => u.Id == id && u.IsDeleted == 0);
        if (user == null) return ApiResponse.Error(404, "用户不存在");
        user.Password = BCrypt.Net.BCrypt.HashPassword(request.NewPassword);
        user.PasswordUpdateTime = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        return ApiResponse.Success("密码重置成功");
    }

    public async Task<ApiResponse<List<RoleBriefDto>>> GetRolesAsync(Guid id)
    {
        var roles = await _db.SysUserRoles
            .Where(ur => ur.UserId == id)
            .Include(ur => ur.Role)
            .Select(ur => new RoleBriefDto { Id = ur.Role.Id, Code = ur.Role.Code, Name = ur.Role.Name })
            .ToListAsync();
        return ApiResponse<List<RoleBriefDto>>.Success(roles);
    }

    public async Task<ApiResponse<List<PermissionBriefDto>>> GetPermissionsAsync(Guid id)
    {
        var perms = await _db.SysUserRoles
            .Where(ur => ur.UserId == id)
            .SelectMany(ur => ur.Role.RolePermissions)
            .Select(rp => rp.Permission)
            .Distinct()
            .Select(p => new PermissionBriefDto { Id = p.Id, Code = p.Code, Name = p.Name })
            .ToListAsync();
        return ApiResponse<List<PermissionBriefDto>>.Success(perms);
    }
}
